## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [889]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [890]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [891]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [892]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [893]:
df.isnull().mean()

first          0.0
last           0.0
job            0.0
phone          0.0
email          0.0
client_id      0.0
order_id       0.0
order_date     0.0
order_week     0.0
order_year     0.0
item_id        0.0
category       0.0
subcategory    0.0
unit_price     0.0
unit_cost      0.0
unit_weight    0.0
qty            0.0
line_number    0.0
dtype: float64

In [894]:
df.duplicated(['order_id'])

0        False
1        False
2        False
3        False
4        False
         ...  
54634     True
54635     True
54636     True
54637     True
54638     True
Length: 54639, dtype: bool

In [895]:
df[54634:54638]

first       last               job               phone  \
54634   Lauren      Reese    Radio producer  977.877.5272x11382   
54635  Derrick      Moore          Musician        358.661.5483   
54636   Monica  Gutierrez  Graphic designer    294.805.9100x339   
54637    Wanda    Solomon      Toxicologist       (311)767-4924   

                                    email  client_id  order_id  order_date  \
54634  late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635        derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636           gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637           solomonwanda5962@ross.org      24485   7592730  2023-01-18   

       order_week  order_year         item_id     category        subcategory  \
54634          17        2023  UCA28532-31-0B     software          logistics   
54635           6        2023  EBD29277-23-7U  consumables   kitchen supplies   
54636           9        2023  EBD30093-55-9Z  consumables  bathroom supplies   
54637           3        2023  XMZ38239-49-3D    furniture             stands   

       unit_price  unit_cost  unit_weight  qty  line_number  
54634       83.13      51.60         2.25   33            8  
54635      206.59     175.46        11.70   47            0  
54636       65.66      57.31         4.16  475            6  
54637        1.48       1.22        18.04  112            0

In [896]:
df['category'].nunique()

5

In [897]:
# What three item categories had the most entries?

df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [898]:
# For the category with the most entries,
# which subcategory had the most entries?

df_s = df.value_counts(['category', 'subcategory'])
df_s.head(1)

category     subcategory      
consumables  bathroom supplies    6424
Name: count, dtype: int64

In [899]:
df_ID = df['client_id'].value_counts().head(5).index
df_ID

Index([33615, 66037, 46820, 38378, 24741], dtype='int64', name='client_id')

In [900]:
# Which five clients had the most entries in the data?
df_ID = df['client_id'].value_counts().head(5)
df_ID

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [901]:
# WHY DIDN'T THIS AGREE WITH THE ABOVE?
group_cust = df.groupby(by='last')['client_id'].count()
top_cust = group_cust.sort_values(ascending=False)
top_cust.head()

last
Johnson      1164
Williams      835
Garcia        687
Jones         632
Hernandez     613
Name: client_id, dtype: int64

In [902]:
# Store the client ids of those top 5 clients in a list.

top_5 = df_ID.index.tolist()

top_5


[33615, 66037, 46820, 38378, 24741]

In [903]:
# How many total units (the qty column) did the
# client with the most entries order order?

df_sum=df.loc[df['client_id'] == top_5[0] , ['qty']].sum()
df_sum

qty    64313
dtype: int64

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [904]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['gross_sales'] = df['unit_price'] * df['qty']
df2=df.loc[:,['unit_price','qty','gross_sales']]
df2.head()


unit_price  qty  gross_sales
0     1096.80  105    115164.00
1       24.95   21       523.95
2       13.52   39       527.28
3       36.42   29      1056.18
4      195.10   20      3902.00

In [905]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def ship_price(weight, qty):
    net_weight = weight * qty
    if net_weight < 50:
        return net_weight * 10
    else:
        return net_weight * 7

df['ship_price'] = df.apply(lambda x: ship_price(x['unit_weight'],x['qty']), axis = 1)
df3=df.loc[:,['unit_price',	'unit_weight',	'qty',	'ship_price']]
df3.head()

unit_price  unit_weight  qty  ship_price
0     1096.80         7.50  105     5512.50
1       24.95         1.49   21      312.90
2       13.52         1.68   39      458.64
3       36.42         1.23   29      356.70
4      195.10        46.43   20     6500.20

In [906]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%



df['total_price'] = (df['gross_sales']+df['ship_price']) * 1.0925
df4=df.loc[:,['gross_sales','ship_price','total_price']]
df4.head()

gross_sales  ship_price    total_price
0    115164.00     5512.50  131839.076250
1       523.95      312.90     914.258625
2       527.28      458.64    1077.117600
3      1056.18      356.70    1543.571400
4      3902.00     6500.20   11364.403500

In [907]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

def cost(qty, unit_cost):
    cogs = qty * unit_cost
    return cogs


df['total_cost'] = df.apply(lambda x: cost(x['qty'], x['unit_cost']), 
                            axis=1) + df['ship_price']


df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  gross_sales  ship_price    total_price  \
0         7.50  105            1    115164.00     5512.50  131839.076250   
1         1.49   21            0       523.95      312.90     914.258625   
2         1.68   39            6       527.28      458.64    1077.117600   
3         1.23   29            3      1056.18      356.70    1543.571400   
4        46.43   20            1      3902.00     6500.20   11364.403500   

   total_cost  
0    85597.05  
1      629.79  
2      765.18  
3     1077.35  
4     8663.60  

[5 rows x 22 columns]

In [908]:
# Create a column for the profit of
# each line using line cost and line price

df['profit'] = df['total_price'] - df['total_cost']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  gross_sales  ship_price    total_price  total_cost  \
0            1    115164.00     5512.50  131839.076250    85597.05   
1            0       523.95      312.90     914.258625      629.79   
2            6       527.28      458.64    1077.117600      765.18   
3            3      1056.18      356.70    1543.571400     1077.35   
4            1      3902.00     6500.20   11364.403500     8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [909]:
# Check your work using the totals above

input_1 = 2742071
input_2 = 2173913
input_3 = 6128929

order_1=df.loc[df['order_id'] == input_1  , ['total_price']].sum()
order_2=df.loc[df['order_id'] == input_2  , ['total_price']].sum()
order_3=df.loc[df['order_id'] == input_3  , ['total_price']].sum()


print(f'For Order ID {input_1} the {order_1} \nFor Order ID {input_2} the {order_2} \nFor Order ID {input_3} the {order_3}')

For Order ID 2742071 the total_price    152811.897075
dtype: float64 
For Order ID 2173913 the total_price    162388.7193
dtype: float64 
For Order ID 6128929 the total_price    923441.24045
dtype: float64


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [910]:
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'gross_sales', 'ship_price', 'total_price', 'total_cost',
       'profit'],
      dtype='object')

In [911]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids 

    # .isin removes all rows that do not intersect the client_id and top_5 list.
    # .groupby groups those intersecting values by the top 5 client IDs.
    # . sum adds values up by client_id
total_spend= df[df['client_id'].isin(top_5)].groupby('client_id')[['gross_sales','ship_price']].sum()
total_spend['Customer Spend']=total_spend['gross_sales']+total_spend['ship_price']
total_spend
    # because the client_id is the index, we can apply a generic format to one column.   
#total_spend.apply(lambda x: "${:,.2f}".format(x))

gross_sales  ship_price  Customer Spend
client_id                                         
24741      70176885.25  5126448.37     75303333.62
33615       5839032.11  1828834.89      7667867.00
38378       8384321.15  3429455.40     11813776.55
46820       7317356.03  1601448.84      8918804.87
66037       7995708.38  1395151.85      9390860.23

In [912]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids 

    # .isin removes all rows that do not intersect the client_id and top_5 list.
    # .groupby groups those intersecting values by the top 5 client IDs.
    # . sum adds values up by client_id
total_spend= df[df['client_id'].isin(top_5)].groupby('client_id')[['total_price']].sum()
#total_spend['Customer Spend']=total_spend['gross_sales']+total_spend['ship_price']
total_spend
    # because the client_id is the index, we can apply a generic format to one column.   
#total_spend.apply(lambda x: "${:,.2f}".format(x))

total_price
client_id              
24741      8.226889e+07
33615      8.377145e+06
38378      1.290655e+07
46820      9.743794e+06
66037      1.025951e+07

In [913]:


 
ts=df[df['client_id'].isin(top_5)].groupby('client_id')['qty'].sum() #finds the total quantity of units purchased for each of the the top 5 customers
ship=df[df['client_id'].isin(top_5)].groupby('client_id')['ship_price'].sum() #finds the total shipping costs inclurreds for each of the top 5 customers
profit=df[df['client_id'].isin(top_5)].groupby('client_id')['profit'].sum()#finds the total profit for each of the top 5 customers
rev=df[df['client_id'].isin(top_5)].groupby(['client_id'])['total_price'].sum()#finds the total profit for each of the top 5 customers

#Extract the index column into a list that will be changed into the first column of a dataframe below.
client5_id_list = df_ID.index.tolist()


# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.


top_5_df =pd.DataFrame(data=client5_id_list, columns=['client_id'])
'''I need to use either a join or merge, using the client ID as the common bond.'''
# Merged the various df into one df.
top_5_df=pd.merge(top_5_df,ts, on='client_id')
top_5_df=pd.merge(top_5_df,ship, on='client_id')
top_5_df=pd.merge(top_5_df,rev, on='client_id')
top_5_df=pd.merge(top_5_df,profit, on='client_id')
#top_5_percent_df=top_5_df['profit']/top_5_df['total_price']
top_5_df




client_id     qty  ship_price   total_price        profit
0      33615   64313  1828834.89  8.377145e+06  2.201981e+06
1      66037   43018  1395151.85  1.025951e+07  3.255032e+06
2      46820   75768  1601448.84  9.743794e+06  2.736603e+06
3      38378   73667  3429455.40  1.290655e+07  3.271830e+06
4      24741  239862  5126448.37  8.226889e+07  3.657999e+07

In [914]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.


def format_comma (x):
    return '{:,.0f}'.format(x)

def format_dollar(x):
    return '${:,.2f}M'.format(x)

def convert_to_int(value):
    try:
        return int(value)
    except (ValueError, TypeError):
        return pd.NA  # Convert non-convertible values to NA

# Changes column titles
top_5_df=top_5_df.rename(columns={"client_id":"Client ID", 'qty':'Units Shipped',"ship_price":'Shipping','total_price':'Total Revenues','profit':'Total Profit'})

# Changes values to millions

top_5_df[['Shipping', 'Total Revenues', 'Total Profit']] = \
    top_5_df[['Shipping', 'Total Revenues', 'Total Profit']]/1000000

# Format $M
top_5_df['Units Shipped'] = top_5_df['Units Shipped'].apply(format_comma)
top_5_df[['Shipping', 'Total Revenues', 'Total Profit']] = \
    top_5_df[['Shipping', 'Total Revenues', 'Total Profit']].applymap(format_dollar)


top_5_df

/var/folders/zb/htjdpgks76b30lkhf98n6s5m0000gn/T/ipykernel_44481/2792321637.py:29: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  top_5_df[['Shipping', 'Total Revenues', 'Total Profit']].applymap(format_dollar)


Client ID Units Shipped Shipping Total Revenues Total Profit
0      33615        64,313   $1.83M         $8.38M       $2.20M
1      66037        43,018   $1.40M        $10.26M       $3.26M
2      46820        75,768   $1.60M         $9.74M       $2.74M
3      38378        73,667   $3.43M        $12.91M       $3.27M
4      24741       239,862   $5.13M        $82.27M      $36.58M

In [915]:
# Sort the updated data by "Total Profit" from highest to lowest
top_5_df= top_5_df.sort_values(["Total Profit"], ascending=False)
top_5_df

Client ID Units Shipped Shipping Total Revenues Total Profit
4      24741       239,862   $5.13M        $82.27M      $36.58M
3      38378        73,667   $3.43M        $12.91M       $3.27M
1      66037        43,018   $1.40M        $10.26M       $3.26M
2      46820        75,768   $1.60M         $9.74M       $2.74M
0      33615        64,313   $1.83M         $8.38M       $2.20M

In [920]:
'''Sumnmary of findings:
 1. While the 2nd - 5th customers, ranked by units shipped, tend to have a profit margin in the 25% range,
     the customer ranked #1 (#24741) has aprofit margin that is significantly higher at 44%.
 2.  The shipping costs per revenue dollar is much lower for the #1 customer (6%) vs. the other customers (~25%).
 3.  The average revenues per unit sold for the #1 client is $320 vs ~$150 for the remaibning top 5.
 
 Conclusion:  This would imply that the items being purhcased by the #1 customer are either a highly value added product 
 (typically typified as having large IP value (e.g..sophisticated semiconductors), something small in short supply and 
 high demand (thus the higher margins) or both.'''


'Sumnmary of findings:\n 1. While the 2nd - 5th customers, ranked by units shipped, tend to have a profit margin in the 25% range,\n     the customer ranked #1 (#24741) has aprofit margin that is significantly higher at 44%.\n 2.  The shipping costs per revenue dollar is much lower for the #1 customer (6%) vs. the other customers (~25%).\n 3.  The average revenues per unit sold for the #1 client is $320 vs ~$150 for the remaibning top 5.\n \n Conclusion:  This would imply that the items being purhcased by the #1 customer are either a highly value added product \n (typically typified as having large IP value (e.g..sophisticated semiconductors), something small in short supply and \n high demand (thus the higher margins) or both.'